crates/symbolicator-service/src/download/mod.rs

```diff
@@ -397,84 +402,96 @@ impl DownloadService {
-    /// downloading you may still need to filter the files.
-    pub async fn list_files(
-        &self,
-        sources: &[SourceConfig],
-        filetypes: &[FileType],
-        object_id: &ObjectId,
-    ) -> Vec<RemoteFile> {
-        let mut remote_files = vec![];
-
-        macro_rules! check_source {
-            ($source:ident => $file_ty:ty) => {{
+    /// Whether this download service is allowed to connect to reserved ip addresses.
+    pub fn can_connect_to_reserved_ips(&self) -> bool {
+        self.connect_to_reserved_ips
+    }
+}
+
+/// Returns all objects matching the [`ObjectId`] at the source.
+///
+/// Some sources, namely all the symbol servers, simply return the locations at which a
+/// download attempt should be made without any guarantee the object is actually there.
+///
+/// If the source needs to be contacted to get matching objects this may fail and
+/// returns a [`CacheError`].
+///
+/// Note that the `filetypes` argument is not more then a hint, not all source types
+/// will respect this and they may return all DIFs matching the `object_id`.  After
+/// downloading you may still need to filter the files.
+pub async fn list_files(
+    downloader: &DownloadService,
+    symstore_index_service: &SymstoreIndexService,
+    sources: &[SourceConfig],
+    filetypes: &[FileType],
+    object_id: &ObjectId,
+) -> Vec<RemoteFile> {
+    let mut remote_files = vec![];
+
+    macro_rules! check_source {
+        ($source:ident => $file_ty:ty) => {{
+            let mut iter =
+                SourceLocationIter::new(&$source.files, filetypes, None, object_id).peekable();
+            if iter.peek().is_none() {
+                // TODO: create a special "no file on source" `RemoteFile`?
+            } else {
+                remote_files.extend(iter.map(|loc| <$file_ty>::new($source.clone(), loc).into()))
+            }
+        }};
+    }
+
+    for source in sources {
+        match source {
+            SourceConfig::Sentry(cfg) => {
+                let future = downloader
+                    .sentry
+                    .list_files(cfg.clone(), object_id, filetypes);
+
+                match future.await {
+                    Ok(files) => remote_files.extend(files),
+                    Err(error) => {
+                        let error: &dyn std::error::Error = &error;
+                        tracing::error!(error, "Failed to fetch file list");
+                        // TODO: create a special "finding files failed" `RemoteFile`?
+                    }
+                }
+            }
+            SourceConfig::Http(cfg) => {
+                // TODO: Obviously don't hardcode this for Intel
+                let index = if cfg.id.as_str() == "sentry:intel" {
+                    symstore_index_service
+                        .fetch_symstore_index(Scope::Global, Arc::clone(cfg))
+                        .await
+                } else {
+                    None
+                };
```

Comment:
Yeah this would need to be part of the source config.

Reply:
See the newest commit.